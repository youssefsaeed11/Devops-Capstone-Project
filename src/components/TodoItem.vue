<template>
  <base-todo-item
    class="TodoItem"
    :item="item"
  >
    <template #checkbox>
      <input
        v-model="completedLocal"
        type="checkbox"
        class="checkbox"
        data-testid="completeCheckbox"
      >
    </template>
    <template #content>
      <div
        v-if="!isEditing"
        class="break-all"
        data-testid="contentEditTrigger"
        @dblclick="enableEdit"
      >
        {{ item.content }}
      </div>
      <input
        v-if="isEditing"
        ref="editInput"
        v-model="contentLocal"
        type="text"
        class="input"
        data-testid="editInput"
        @keyup.enter="saveEdit"
        @blur="saveEdit"
        @keyup.esc="stopEdit"
      >
    </template>
    <template #actions>
      <button
        class="TodoItem__delete button is-inverted is-danger is-small"
        data-testid="deleteTrigger"
        @click.prevent="deleteItem"
      >
        <span class="icon"><i class="fa fa-trash fa-lg" /></span>
      </button>
      <div class="TodoItem__reorderHandle">
        <i class="fa fa-bars" />
      </div>
    </template>
  </base-todo-item>
</template>

<script>
import BaseTodoItem from '@/components/BaseTodoItem'

/**
 * @module TodoItem
 */
export default {
  name: 'TodoItem',
  components: { BaseTodoItem },
  props: {
    /**
     *  @type TodoItem
     */
    item: {
      type: Object,
      required: true
    }
  },
  data () {
    return {
      isEditing: false,
      contentLocal: ''
    }
  },
  computed: {
    completedLocal: {
      /**
       * Transforms the completedAt timestamp to boolean
       */
      get () {
        return !!this.item.completedAt
      },
      /**
       * Converts the completedAt boolean to a timestamp
       * Emits the entire changed item
       * @param {Boolean} value
       */
      set (value) {
        this.emitChange({
          completedAt: value ? Date.now() : null
        })
      }
    }
  },
  methods: {
    /**
     * Deletes a todo item
     */
    deleteItem () {
      this.emitChange({
        deletedAt: Date.now()
      })
    },
    /**
     * Emits the changes to the TODO item up to the parent.
     * @param {Object} updatedItem - the todo item
     */
    emitChange (updatedItem) {
      this.$emit('change', {
        ...this.item,
        ...updatedItem
      })
    },
    /**
     * Starts the editing of a todo item.
     * Focuses the input item immediately
     * @return {Promise<void>}
     */
    async enableEdit () {
      this.contentLocal = this.item.content
      this.isEditing = true
      await this.$nextTick()
      this.$refs.editInput.focus()
    },
    /**
     * Saves the editing changes.
     * If no content is provided, just closes
     */
    saveEdit () {
      if (this.contentLocal) {
        this.emitChange({
          content: this.contentLocal
        })
      }
      this.stopEdit()
    },
    /**
     * Stops the editing, without saving anything
     */
    stopEdit () {
      this.isEditing = false
      this.contentLocal = ''
    }
  }
}
</script>

<style lang='scss' scoped>
.TodoItem {
  &:hover {
    .TodoItem__delete {
      opacity: 1;
    }
  }

  &__delete {
    opacity: 0;
    margin-right: $theme--spacing-s;
    border-radius: 100%;
  }

  &__reorderHandle {
    &:hover {
      cursor: grab;
    }

    &:active {
      cursor: grabbing;
    }
  }
}
</style>
