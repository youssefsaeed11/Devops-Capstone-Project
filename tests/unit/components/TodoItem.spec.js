import TodoItem from '@/components/TodoItem'
import { completedItem, notDeletedItem } from '../data/ToDoFactory'
import { mount } from '@vue/test-utils'

/**
 * @param overrides
 * @return {{wrapper: Wrapper<Vue>, config: any}}
 */
const createWrapper = (overrides) => {
  let config = Object.assign(
    {
      propsData: {
        item: notDeletedItem()
      }
    },
    overrides
  )

  const wrapper = mount(TodoItem, config)

  return {
    config,
    wrapper
  }
}

describe('TodoItem', () => {
  it('renders the TodoItem', () => {
    const { wrapper } = createWrapper()
    expect(wrapper).toMatchSnapshot()
  })

  it('renders a functioning checkbox', () => {
    const { wrapper } = createWrapper()
    let checkbox = wrapper.find(testid('completeCheckbox'))
    expect(checkbox.exists()).toBeTruthy()
    expect(checkbox.element.checked).toBe(false)
    wrapper.setProps({
      item: completedItem()
    })
    checkbox = wrapper.find(testid('completeCheckbox'))
    expect(checkbox.element.checked).toBe(true)
  })

  it('renders the content if not editing', async () => {
    const { wrapper } = createWrapper()
    wrapper.find(testid('completeCheckbox')).setChecked()
    await wrapper.vm.$nextTick()
    expect(wrapper.emitted('change')[0][0]).toMatchObject({
      completedAt: expect.any(Number)
    })
  })

  it('allows editing the input on content double click', async () => {
    const { wrapper } = createWrapper()
    expect(wrapper.find(testid('editInput')).exists()).toBe(false)

    wrapper.find(testid('contentEditTrigger')).trigger('dblclick')

    await wrapper.vm.$nextTick()
    expect(wrapper.find(testid('editInput')).exists()).toBe(true)
  })

  it('sets the todo item content as a value of the input', () => {
    const { wrapper } = createWrapper()
    wrapper.find(testid('contentEditTrigger')).trigger('dblclick')

    expect(wrapper.find(testid('editInput')).element.value).toBe('notDeleted')
  })

  it('does not emit an event if no content in the edit', async () => {
    const { wrapper } = createWrapper()
    wrapper.find(testid('contentEditTrigger')).trigger('dblclick')
    await wrapper.vm.$nextTick()
    wrapper.find(testid('editInput')).setValue('')
    wrapper.find(testid('editInput')).trigger('keyup.enter')
    await wrapper.vm.$nextTick()
    expect(wrapper.emitted('change')).toBeFalsy()
  })

  it('reverts back when editing is finished', async () => {
    const { wrapper } = createWrapper()
    wrapper.find(testid('contentEditTrigger')).trigger('dblclick')
    await wrapper.vm.$nextTick()
    wrapper.find(testid('editInput')).setValue('New content')
    wrapper.find(testid('editInput')).trigger('keyup.enter')
    await wrapper.vm.$nextTick()
    expect(wrapper.emitted('change')[0][0]).toMatchObject({
      content: 'New content'
    })
    expect(wrapper.find(testid('editInput')).exists()).toBe(false)
  })

  it('sets the item as deleted on deleted click', async () => {
    const { wrapper } = createWrapper()
    wrapper.find(testid('deleteTrigger')).trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper.emitted('change')[0][0]).toMatchObject({
      deletedAt: expect.any(Number)
    })
  })
})
