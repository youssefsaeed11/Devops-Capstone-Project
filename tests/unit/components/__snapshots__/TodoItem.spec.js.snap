// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem renders the TodoItem 1`] = `
<div class="BaseTodoItem TodoItem">
  <div class="BaseTodoItem__checkbox"><input type="checkbox" data-testid="completeCheckbox" class="checkbox"></div>
  <div class="BaseTodoItem__content">
    <div data-testid="contentEditTrigger" class="break-all">
      notDeleted
    </div>
    <!---->
  </div>
  <div class="BaseTodoItem__actions"><button data-testid="deleteTrigger" class="TodoItem__delete button is-inverted is-danger is-small"><span class="icon"><i class="fa fa-trash fa-lg"></i></span></button>
    <div class="TodoItem__reorderHandle"><i class="fa fa-bars"></i></div>
  </div>
</div>
`;
